API: add core.h header file for main public API

garlick opened the pull request:
This PR is a continuation of #173, in which we can discuss a revamp of the Flux public API.
I will add commits as we go and squash them down at the end.  Since in-tree code is built against `src/include/flux/core.h` and `src/lib/libcore/core.h` is only for external users, I modified the latter in this PR.

* I think we agreed in #173 that there should be one `flux_open()`/`flux_close()` which can internally select the appropriate back end handle implementation.  I renamed the open flags to `FLUX_O_*` (from `FLUX_FLAGS_*`) to make it clear they are open flags.

* Pulled in the low-level `flux_sendmsg()`, `flux_recvmsg()`, `flux_putmsg()` functions.  I changed the prototypes slightly so they all look similar, with `FLUX_IO_*` flags expressing the optional behavior.  Converted to flux-specific opaque message type `flux_msg_t` as we had discussed.

* Pulled in the high level `flux_publish()`, `flux_request()`, and `flux_rpc()` functions.  JSON arguments are now valid json strings as we had discussed.  I added a timeout to `flux_rpc()`.

grondo commented:
This seems like a good start, I can't really see any fault in it.

Should we perhaps be explicit in naming the `path` arg to `flux_open()` as a URI?

I'm not exactly sure the mechanism that will be used to translate a call into the the generic `flux_open()` from `libflux.so` into a specific open handler. Perhaps we need to require that the `uri` argument is a [rfc1630](https://tools.ietf.org/html/rfc1630) valid URI, and the libflux code will parse the URI and use the URI *scheme* to load a specific implementation?

Also, is the planned scheme for extending this core api is to have modules or other tools built on flux install a header file under `*sysconfdir*/flux` and a corresponding `flux-**module**` library that implements the new API functions? (I'm fine with this, but just getting us all on the same page)

Perhaps we need to think of a few use cases, and API design will follow, e.g., so far we have

 * in-process module (handle implied)
 * module-as-a-process (call `flux_open`, register a service, then enter `main()` loop)
 * Strict API user (e.g. `flux-*` utilities)
 * Third-party bindings (e.g. @trws Cython bindings)
 * New `open` implementations, e.g. KRB5 connection to `flux-broker` from remote system? (How are these distributed and/or implemented?)

We also want to be able to easily handle 3rd party providers distributing modules out-of-tree...

Any more? Maybe Dong can provide some possible tools scenarios?

trws commented:
It looks good to me as well.  Having it dispatch based on a URI is probably reasonable, maybe with an assumable default fall-back on a local connection if the URI is incomplete or unspecified?

The tricky part of this, as far as I've been thinking, is going to be dealing with the split between library and module interfaces to things like the KVS.  As it stands, the module provides the functionality, and any code that uses the KVS links directly to the library for the interface, correct?  

Jim mentioned at last week's meeting a desire to load more than one version of a module, and this type of set-up is likely to cause headaches for that use-case.  The question it brings up for me is how to define interfaces between flux components or modules in addition to the flux interface itself.  For example, is it accessed directly by users, is it accessed through flux messages only, or could it be accessed by requesting an interface handle of some sort from flux?

I'm probably getting too far in the weeds for this PR, but how we handle the resolution of non-core components I think is going to be an important part of the API design once we're past these fundamental flux calls.  For some reason I'm tempted to suggest offering a way to "import" a module through a flux call, and receive an accessor handle to the module that way, to keep flux in control of what version of the thing is being used by the callee, but that might be complete overkill.

garlick commented:
@grondo: On `flux_open()`, maybe it's a good idea if we ignore the existing zeromq URI schemes and implement our own that map in a deterministic way to a shared library containing the implementation?  For example, (ignoring the fact that "api" is a lame name): `api:///tmp//my/socket` versus `k5tcp:///host:port:principal` or something like that, mapping to `libapi.so` and `libk5tcp.so`?

Your description of extensions is exactly what I had in mind.  It seems like that needs to land in an RFC at some point.

Good idea on the use cases.  I guess I was focusing on the "strict API user" thus far.

grondo commented:
@garlick, yeah that is what I was thinking of by using the URI *scheme* name to load an implementation. @trws is right on with his suggestion of using a default when no URI scheme is given.

@trws, I like your idea of an "import-like" command for loading module implementations.

Another use case for loading modules under different names might be to load the *same* module under a different name... maybe with different options, or perhaps for a short term use. We should make sure we can handle this case as well. (One example of this I've been thinking of is loading a second implementation of the KVS as short term storage, e.g. a user *mounts* a new kvs instance as "tmpdata" then uses it to write a bunch of shared data, then unloads the kvs instance after some post processing, freeing all memory. Maybe this is contrived?)

Ugh. That might have been off-topic for this PR...

garlick commented:
The import-like command seems like a really good idea.  Let's take that one off to issue #176

trws commented:
Good idea.   A couple of things just occurred to me though.

Everything in here is good as it is, but there isn't an interface to create, manipulate or read a flux_msg_t yet.  We probably need equivalents to most of the functions in [messages.h](https://github.com/flux-framework/flux-core/blob/0d021bc691862140a9c44431fd8521ac01d7068b/src/common/libflux/message.h#L1) down through flux_msg_cmp right? 

I'm also thinking a `size_t flux_json_decode(const char *json_str, void** json_bin)` or similar option might be a good idea, but it might just be me.  The idea would be to make it possible for users to convert a string into flux's preferred json binary representation once, then re-use it to break down the repeated encode/decode cost where appropriate, but maybe just being able to create a message from it with something like `flux_msg_set_payload_json` is enough.

garlick commented:
@trws: regarding `message.h`, I suspect we will probably need much of it, but I was going to hold off a bit and work through the simpler use cases first to reach consensus on the most important interfaces.

Regarding the json argument, since messages have a separate "protocol block" used by the broker, the json payload shouldn't need to be looked at by any layers between sender and recipient, and so I am not sure there is any drawback to the public API dealing exclusively with json strings, e.g.:
* protocol codec: values to json string
* public API: send json string
* (multi-hop routing)
* public API: recv json string
* protocol codec: json string to values

In fact, the codecs (some of which are be included in the public API) shouldn't need to expose the encoded json object either.  When they do we can wrap the object, e.g. take the codec for `cmb.lsmod`:
```
json_object *flux_lsmod_json_create (void);
int flux_lsmod_json_append (json_object *a, const char *name, int size,
                            const char *digest, int idle);
```
here the encoded json object is not created in a single encode operation.  So if instead we had:
```
lsmod_t flux_lsmod_create (void);
int flux_lsmod_add (lsmod_t obj, const char *name, int size, const char *digest, int idle);
char *flux_lsmod_encode (lsmod_t obj);
void flux_lsmod_destroy (lsmod_t obj);

```
Am I making any sense?

trws commented:
@garlick You are certainly making sense.  Indirections like that make much more sense than tying it to a binary json representation of some kind.

As to starting with the simpler use-cases, I agree that it's a good idea, but the send, recv and put functions all depend on an un-specified opaque data type.  Should we hold off on those for now as well, or perhaps just rely on the publish, request, and rpc interfaces for the time being?  Come to think of it, for those to work we would at least need flux_msghandler_add as well wouldn't we, or at least some receive-side equivalent?

garlick commented:
I was having another thought that might appease you temporarily :-)  Instead of `flux_publish()` and `flux_request()`, does it make sense to split out the message creation, send, and destroy? E.g.
```
flux_msg_t flux_msg_create_request (uint32_t nodeid, const char *topic, 
                                                             const char *json_in);
int flux_sendmsg (flux_t h, flux_msg_t msg, int flags);
void flux_msg_destroy (flux_msg_t msg);
```
and
```
flux_msg_t flux_msg_create_event (const char *topic, const char *json_in);
int flux_sendmsg (flux_t h, flux_msg_t msg, int flags);
void flux_msg_destroy (flux_msg_t msg);
```
That way, for example, if the send fails with a transient error, you don't have to recreate the message to try again.  For that matter maybe `flux_rpc()` should be broken up too...

`flux_msghandler_add()` belongs to the reactor interface and that whole thing needs work for sure, but first lets work through the functions needed by the simplest utility.  (In case it is unclear, the reactor is not necessary for `flux_rpc()` to work.  `flux_rpc()`  calls `flux_recvmsg()` in a loop waiting for its response, then calls `flux_putmsg()` to put back any other messages recveived while it was waiting.  This synchronous behavior is anathema to the reactor environment which is why we added coprocs...

Thoughts?

trws commented:
That would work for me.  I'm curious about the request nodeid though, why have that on the request create rather than on the send?  It seems a bit backwards for some reason.

Fair point about the reactor, it's just that I'm not sure `flux_subscribe` is worthwhile without a way to handle an incoming message that conforms to the topic.  Specifying a handler in the subscribe call would do it too, I suppose, but that would require larger changes to the reactor would it not?

garlick commented:
I've had similar thoughts about `nodeid`.  It's part of the message so I've included it in message manipulation functions, and it's only relevant for messages of type `FLUX_MSGTYPE_REQUEST`, hence is not included in `flux_sendmsg()` which accepts any message type.

The subscribe affects what events will be seen via `flux_recvmsg()`.  Maybe it's time to pull in `flux_recvmsg_match()` and dicsuss.  Patch on the way...

trws commented:
That makes sense.  I would probably lean more toward treating events and directed messages as different, much the way publish and request already are, rather than making it a message property but that's probably just personal bias.

I see what you mean now.  For some reason it hadn't connected in my brain that flux_recvmsg would take in event messages as well as requests.  

Given that, a minimal example is probably possible, though a non-trivial one would probably need a way to get the topic and payload out of a received message.  I guess the RPC version takes care of that for the caller, but would this be enough for the callee?

garlick commented:
I wasn't taking into consideration the callee at the moment.

I'm beginning to think you were right that we need to bring in `message.h` and all those access functions.

trws commented:
Works for me.
Probably a stupid question, but is the routing frame material something that the user should see?  I need to get a better handle on how all of that is set up, but it seems like something that would be in an "advanced" interface of some kind.

garlick commented:
Dropped, and I added an accessor for the sender of a request.  We can always pull them in again if necessary.

dongahn commented:
Sorry, I got distracted by TotalView user issues. 

@grondo Re: Any more? Maybe Dong can provide some possible tools scenarios?

I believe you have already well covered the cases for tools! It is good that we now have multiple design choices to support run-time tools like LaunchMON-based tools. 

From my earlier prototyping, I know I can make the LaunchMON engine a strict API user, which simply `flux_open` and *uses* Flux’s services (e.g., kvs and TP execution services) to fetch information about the target job (e.g., proctable) and co-locate the tool daemons with the target MPI processes. So this will be likely my initial path again. 

I was also considering whether it is a good idea to embed the launchmon engine directly as a module (-as-a-process) and thinking how much efforts this will need. The advantages of this *persistent service" would be to save its client tool from having to launch an engine for every tool session. 

But at first glance, this approach seems to require too much efforts on my part. In case you're interested, what is going on in my mind is that:

1) With this module approach, we will need a more rigorous mechanism to control protocol match/mismatch between the launchmon engine and launchmon's tool users (e.g., STAT). Fortunately, this probably can be easily handled if I port its native LMONP communication protocol to google protobuf. 

2) A slightly bigger issue would be on how to wrap its TCP based service into flux messages. It just seems wrong if the launchmon engine gets loaded and then opens up a TCP port and starts to service through this communication mechanism (different from the flux messaging system). If it does, the only advantage of the module approach would be the flux instance forking/execing this persistent service for us. So to do this right, I think we will have to wrap LMONP TCP-based service in the flux messages and rely on the `comms` routing . Without prototyping, I don't know what it will take us to do this.  BTW, this may be a general question for module developers -- when they have an existing service, what would be the best practice to port them to our `cmdb` scheme... My previous works with MRNet suggests that wrapping an application level protocol as a payload of the communication infrastructure is a reasonable approach But I'm not sure `cmbd` also supports or encourage this.

3) This is more of a design issue with LaunchMON itself. But LaunchMON currently requires a separate launchmon engine process per tool session (i.e., since ptrace only supports one debugger process to trace debugger process(es) and the engine uses ptrace to trace the MPI starter process, this design was chosen.) But with the engine as a module, we will probably need to make one module instantiation serve multiple tool sessions … 

This is a long way to say I will still have to think this through. 

What I might end up doing could simply make the launchmon engine a strict API user. Then, I may add an additional "launchmon assisting module" to boost certain LaunchMON's operations (and thus tool operations). But until I crack open this project, it is hard to tell what's the best. 

Nevertheless,  a module as a process does offer good design choices for tools!

dongahn commented:
@grondo Re: We also want to be able to easily handle 3rd party providers distributing modules out-of-tree...

The scheduler framework service module would an example for this and I am willing to be a tester for this mechanism :-) Just $.02

dongahn commented:
Speaking of use cases, another important class of strict API users will be MPI libraries and other distributed software infrastructures (e.g., MRNet). I have to think it will be rare that an MPI implementation or a distributed SW infrastructure will directly interface with Flux API. But they will be  layered with some known bootstrapping APIs including PMI. So we just want to make sure these external boostrapping APIs can be implemented using the pubic API set. (I see src/lib/libpmi already has a PMI implementation but there will be 3rd party providers who will want to build other similar boostrapping libraries.)

I have to think they will mostly want to use KVS and barriers... Did we plan to discuss the public APIs for those services or limit the discussions only to core functionalities  as part of this PR? 

Best,
Dong

garlick commented:
In a way I kind of feel like we've failed if tools (or application bootstrap environments) need to set up their own overlay network.  It probably makes sense in extreme cases where the tool needs particular topology or fanout, but it seems like for the common case we should enable rapid and uncomplicated tool launch.  Maybe it is worth studying the launchmon case in detail to see how it would map to our API.  I would suggest that we open a separate issue to discuss launchmon.  I can do that Dong if you agree that it is worth at least understanding any mismatch.

dongahn commented:
Sounds good. Please open a separate issue to discuss this.

dongahn commented:
I just read some of the discussions. Tom is making really good points. BTW, if message manipulation functions will be ultimately added, do we really need separate message creation routines, one for request and the other event? Or having the user construct a right type of message through a msg manipulation API and then send, destroy. 

Ether way is fine by me. But it seems it would be good to understand the design philosophy of these public APIs. Is the intent to provide 

- "high level" APIs
- "medium level" APIs
- "low level" advanced APIs?

Or simply identify commonly used functionality and provide only one set of API?  Either is fine, but understanding/setting an approach might allow us to look at the APIs (as being brought in) with specific angles.

garlick commented:
@dongahn: I opened issue #177 for launchmon.

garlick commented:
I'm not sure we've created the right structure for covering all the ground we need to here.
If you don't mind I'd like to regroup (again, sorry).

* we've split off the launchmon discussion to #177.
* we've split off the module handle discussion to #176.
* I will add a new issue for the generic `flux_open()` under #170 tracker.
* I will add a new issue to pull module interfaces out of `libflux-core.so` to resolve the circular linking problems that led us to statically link everything in-tree (related to #176 but solvable independently) also under #170 tracker
* I think the use cases are probably the best way to decide what should be in the public vs private API's so I will add those to #172 and try to capture some of the discussion related to that from this PR
* The reactor interface needs to be revisited.  I will link #124 to the #170 tracker

garlick commented:
Closing PR.  Discussion on related topics may continue in the issues referenced above.
